c2pa_c_ffi/src/c_api.rs

```diff
@@ -895,6 +895,86 @@ pub unsafe extern "C" fn c2pa_builder_add_ingredient_from_stream(
     ok_or_return_int!(result, |_| 0) // returns 0 on success
 }
 
+/// Adds an action to the manifest the Builder is constructing.
+///
+/// # Parameters
+/// * builder_ptr: pointer to a Builder.
+/// * action_json: JSON string containing the action data.
+///
+/// # Errors
+/// Returns -1 if there were errors, otherwise returns 0.
+/// The error string can be retrieved by calling [c2pa_error].
+///
+/// # Safety
+/// Reads from NULL-terminated C strings.
+///
+/// # Example
+/// ```C
+/// const char* manifest_def = "{}";
+/// C2paBuilder* builder = c2pa_builder_from_json(manifest_def);
+///
+/// const char* action_json = "{\n"
+///     "    \"action\": \"com.example.test-action\",\n"
+///     "    \"parameters\": {\n"
+///     "        \"key1\": \"value1\",\n"
+///     "        \"key2\": \"value2\"\n"
+///     "    }\n"
+/// "}";
+///
+/// int result = c2pa_builder_add_action(builder, action_json);
+/// ```
+///
+/// This creates a manifest with an actions assertion
+/// containing the added action (excerpt of the full manifest):
+/// ```json
+/// "assertions": [
+///   {
+///     "label": "c2pa.actions.v2",
+///     "data": {
+///       "actions": [
+///         {
+///           "action": "c2pa.created",
+///           "digitalSourceType": "http://c2pa.org/digitalsourcetype/empty"
+///         },
+///         {
+///           "action": "com.example.test-action",
+///           "parameters": {
+///             "key2": "value2",
+///             "key1": "value1"
+///           }
+///         }
+///       ],
+///     }
+///   }
+/// ]
+/// ```
+#[no_mangle]
+pub unsafe extern "C" fn c2pa_builder_add_action(
+    builder_ptr: *mut C2paBuilder,
+    action_json: *const c_char,
+) -> c_int {
+    check_or_return_int!(builder_ptr);
```

Comment:
guard_boxed_int!() includes the null check so you don't need to do this check first.